server/src/main/java/org/elasticsearch/common/bytes/CompositeBytesReference.java

```diff
@@ -164,6 +164,26 @@ public BytesReference slice(int from, int length) {
         return CompositeBytesReference.ofMultiple(inSlice);
     }
 
+    @Override
+    public BytesReference copy(int from, int length) {
+        Objects.checkFromIndexSize(from, length, this.length);
+        final int to = from + length;
+        final int limit = getOffsetIndex(to - 1);
+        final int start = getOffsetIndex(from);
+        final int numCopies = 1 + (limit - start);
+        final int inCopyOffset = from - offsets[start];
+        if (numCopies == 1) {
+            return references[start].copy(inCopyOffset, length);
+        }
+        final BytesReference[] inCopy = new BytesReference[numCopies];
```

Comment:
If we wanted a single humongous `byte[]`, we could use `BytesReference.toBytesRef()`. The point is that we are trying to avoid that, because it turns out to be fairly hostile to the G1GC implementation in recent JDKs.

Reply:
Ah I see you've thought about it so fine :)